File: src/main/java/mediathek/gui/duplicates/overview/FilmDuplicateOverviewDialog.java
/*
 * Created by JFormDesigner on Wed Oct 23 21:39:11 CEST 2024
 */

package mediathek.gui.duplicates.overview;

import ca.odell.glazedlists.BasicEventList;
import ca.odell.glazedlists.EventList;
import ca.odell.glazedlists.swing.GlazedListsSwing;
import mediathek.config.Daten;
import mediathek.daten.DatenFilm;
import mediathek.gui.duplicates.details.DuplicateFilmDetailsTableFormat;
import mediathek.tool.EscapeKeyHandler;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreeSelectionModel;
import java.awt.*;
import java.util.Comparator;
import java.util.List;

/**
 * @author christianfranzke
 */
public class FilmDuplicateOverviewDialog extends JDialog {

    private final EventList<DatenFilm> filmList = new BasicEventList<>();

    public FilmDuplicateOverviewDialog(Window owner) {
        super(owner);
        initComponents();
        EscapeKeyHandler.installHandler(this, this::dispose);

        okButton.addActionListener(e -> dispose());
        //must be called for tooltips working
        ToolTipManager.sharedInstance().registerComponent(tree);

        var rootNode = new DefaultMutableTreeNode("Filmduplikate", true);
        var senderList = getSenderList();
        for (var sender : senderList) {
            var node = getDuplicatesForSender(sender);
            //skip empty nodes
            if (node.getChildCount() > 0)
                rootNode.add(node);
        }

        var model = GlazedListsSwing.eventTableModelWithThreadProxyList(filmList, new DuplicateFilmDetailsTableFormat());
        table.setModel(model);
        resetColumnWidths();

        var treeModel = new DefaultTreeModel(rootNode);
        tree.setModel(treeModel);
        var selectionModel = tree.getSelectionModel();
        selectionModel.setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);
        tree.setCellRenderer(new CustomTreeCellRenderer());
        selectionModel.addTreeSelectionListener(e -> {
            var node = (DefaultMutableTreeNode)tree.getLastSelectedPathComponent();
            if (node == null)
                return;
            if (node.getUserObject() instanceof DatenFilm f) {
                var list = Daten.getInstance().getListeFilme()
                        .parallelStream()
                        .filter(item -> !item.isLivestream())
                        .filter(item -> item.getUrlNormalQuality().equals(f.getUrlNormalQuality())
                                        && item.getHighQualityUrl().equals(f.getHighQualityUrl()))
                        .toList();

                filmList.clear();
                filmList.addAll(list);
                calculateColumnWidths();
            }
            else {
                // just clear table
                filmList.clear();
                resetColumnWidths();
            }
        });
    }

    private void calculateColumnWidths() {
        table.getColumnModel().getColumns().asIterator().forEachRemaining(column -> {
            int preferredWidth = column.getMinWidth();
            int maxWidth;
            int colIdx = column.getModelIndex();
            var defaultHeaderRenderer = table.getTableHeader().getDefaultRenderer();
            var columnHeaderRenderer = column.getHeaderRenderer();
            if (columnHeaderRenderer == null)
                columnHeaderRenderer = defaultHeaderRenderer;
            var header = columnHeaderRenderer.getTableCellRendererComponent(table, column.getHeaderValue(), false, false, 0, colIdx);
            maxWidth = header.getPreferredSize().width;

            for (int row = 0; row < table.getRowCount(); row++){
                var cellRenderer = table.getCellRenderer(row, colIdx);
                var c = table.prepareRenderer(cellRenderer, row, colIdx);
                int width = c.getPreferredSize().width + table.getIntercellSpacing().width;
                preferredWidth = Math.max(preferredWidth, width);

                //  We've exceeded the maximum width, no need to check other rows
                if (preferredWidth <= maxWidth){
                    preferredWidth = maxWidth;
                    break;
                }
            }
            column.setPreferredWidth(preferredWidth);
        });

        //update must be manually triggered after width modification
        table.doLayout();
    }

    private void resetColumnWidths() {
        table.getColumnModel().getColumns().asIterator().forEachRemaining(column -> column.setPreferredWidth(90));
        table.doLayout();
    }

    private DefaultMutableTreeNode getDuplicatesForSender(String sender)
    {
        var root = new DefaultMutableTreeNode(sender);

        var list = Daten.getInstance().getListeFilme().parallelStream()
                .filter(DatenFilm::isDuplicate)
                .filter(f -> f.getSender().equals(sender))
                .sorted(Comparator.comparing(DatenFilm::getTitle))
                .toList();
        //System.out.println("List size: " + list.size() + " for sender: " + sender);
        for (var f : list) {
            var node = new DefaultMutableTreeNode(f);
            root.add(node);
        }

        return root;
    }

    private List<String> getSenderList() {
        // "ARD" und "ZDF" immer am Ende um die kleineren Mediatheken nicht zu benachteiligen
        // Alphabetisch sortieren für alle anderen

        return Daten.getInstance().getListeFilme().parallelStream()
                .map(DatenFilm::getSender)
                .distinct()
                .sorted((o1, o2) -> {
                    // "ARD" und "ZDF" immer am Ende um die kleineren Mediatheken nicht zu benachteiligen
                    if (o1.equals("ARD") || o1.equals("ZDF")) {
                        return 1;
                    }
                    if (o2.equals("ARD") || o2.equals("ZDF")) {
                        return -1;
                    }
                    // Alphabetisch sortieren für alle anderen
                    return o1.compareTo(o2);
                })
                .toList();
    }

    private void initComponents() {
        // JFormDesigner - Component initialization - DO NOT MODIFY  //GEN-BEGIN:initComponents  @formatter:off
        // Generated using JFormDesigner non-commercial license
        var dialogPane = new JPanel();
        var contentPanel = new JPanel();
        var splitPane1 = new JSplitPane();
        var scrollPane1 = new JScrollPane();
        tree = new JTree();
        var scrollPane2 = new JScrollPane();
        table = new JTable();
        var buttonBar = new JPanel();
        okButton = new JButton();

        //======== this ========
        setTitle("\u00dcbersicht aller Duplikate"); //NON-NLS
        setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
        setModal(true);
        setPreferredSize(new Dimension(640, 480));
        var contentPane = getContentPane();
        contentPane.setLayout(new BorderLayout());

        //======== dialogPane ========
        {
            dialogPane.setBorder(new EmptyBorder(12, 12, 12, 12));
            dialogPane.setPreferredSize(new Dimension(800, 600));
            dialogPane.setLayout(new BorderLayout());

            //======== contentPanel ========
            {
                contentPanel.setLayout(new BorderLayout());

                //======== splitPane1 ========
                {
                    splitPane1.setDividerLocation(350);

                    //======== scrollPane1 ========
                    {
                        scrollPane1.setViewportView(tree);
                    }
                    splitPane1.setLeftComponent(scrollPane1);

                    //======== scrollPane2 ========
                    {

                        //---- table ----
                        table.setAutoResizeMode(JTable.AUTO_RESIZE_OFF);
                        scrollPane2.setViewportView(table);
                    }
                    splitPane1.setRightComponent(scrollPane2);
                }
                contentPanel.add(splitPane1, BorderLayout.CENTER);
            }
            dialogPane.add(contentPanel, BorderLayout.CENTER);

            //======== buttonBar ========
            {
                buttonBar.setBorder(new EmptyBorder(12, 0, 0, 0));
                buttonBar.setLayout(new GridBagLayout());
                ((GridBagLayout)buttonBar.getLayout()).columnWidths = new int[] {0, 80};
                ((GridBagLayout)buttonBar.getLayout()).columnWeights = new double[] {1.0, 0.0};

                //---- okButton ----
                okButton.setText("Schlie\u00dfen"); //NON-NLS
                buttonBar.add(okButton, new GridBagConstraints(1, 0, 1, 1, 0.0, 0.0,
                    GridBagConstraints.CENTER, GridBagConstraints.BOTH,
                    new Insets(0, 0, 0, 0), 0, 0));
            }
            dialogPane.add(buttonBar, BorderLayout.SOUTH);
        }
        contentPane.add(dialogPane, BorderLayout.CENTER);
        pack();
        setLocationRelativeTo(getOwner());
        // JFormDesigner - End of component initialization  //GEN-END:initComponents  @formatter:on
    }

    // JFormDesigner - Variables declaration - DO NOT MODIFY  //GEN-BEGIN:variables  @formatter:off
    // Generated using JFormDesigner non-commercial license
    private JTree tree;
    private JTable table;
    private JButton okButton;
    // JFormDesigner - End of variables declaration  //GEN-END:variables  @formatter:on
}

File: src/main/kotlin/mediathek/gui/duplicates/overview/FilmDuplicateOverviewDialog.kt
/*
 * Created by JFormDesigner on Wed Oct 23 21:39:11 CEST 2024
 */

package mediathek.gui.duplicates.overview

import ca.odell.glazedlists.BasicEventList
import ca.odell.glazedlists.EventList
import ca.odell.glazedlists.swing.GlazedListsSwing
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import kotlinx.coroutines.withContext
import mediathek.config.Daten
import mediathek.daten.DatenFilm
import mediathek.gui.duplicates.details.DuplicateFilmDetailsTableFormat
import mediathek.tool.EscapeKeyHandler
import java.awt.Window
import javax.swing.ToolTipManager
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.TreeSelectionModel

class FilmDuplicateOverviewDialog(owner: Window) : FilmDuplicateOverviewDialogBase(owner) {
    private val dialogScope = CoroutineScope(SupervisorJob() + Dispatchers.Swing)
    private val filmList: EventList<DatenFilm> = BasicEventList()
    private var selectionJob: Job? = null

    init {
        EscapeKeyHandler.installHandler(this, this::dispose)
        okButton.addActionListener { dispose() }
        setupTable()
        setupTree()
        loadDuplicateTree()
    }

    override fun dispose() {
        dialogScope.cancel()
        super.dispose()
    }

    private fun setupTable() {
        table.model = GlazedListsSwing.eventTableModelWithThreadProxyList(
            filmList,
            DuplicateFilmDetailsTableFormat(),
        )
        resetColumnWidths()
    }

    private fun setupTree() {
        ToolTipManager.sharedInstance().registerComponent(tree)
        tree.selectionModel.selectionMode = TreeSelectionModel.SINGLE_TREE_SELECTION
        tree.cellRenderer = CustomTreeCellRenderer()
        tree.addTreeSelectionListener { updateSelectedDuplicateFilms() }
    }

    private fun loadDuplicateTree() {
        dialogScope.launch {
            val rootNode = withContext(Dispatchers.Default) {
                createDuplicateRootNode(Daten.getInstance().listeFilme.snapshot())
            }
            tree.model = DefaultTreeModel(rootNode)
        }
    }

    private fun createDuplicateRootNode(films: List<DatenFilm>): DefaultMutableTreeNode {
        val rootNode = DefaultMutableTreeNode(ROOT_NODE_LABEL, true)
        val duplicateFilmsBySender = films.asSequence()
            .filter { film -> film.isDuplicate }
            .groupBy { film -> film.sender }

        duplicateFilmsBySender.keys
            .sortedWith(::compareSenders)
            .forEach { sender ->
                val senderNode = DefaultMutableTreeNode(sender)
                duplicateFilmsBySender.getValue(sender)
                    .sortedBy { film -> film.title }
                    .forEach { film -> senderNode.add(DefaultMutableTreeNode(film)) }
                rootNode.add(senderNode)
            }

        return rootNode
    }

    private fun updateSelectedDuplicateFilms() {
        val node = tree.lastSelectedPathComponent as? DefaultMutableTreeNode ?: return
        val film = node.userObject as? DatenFilm
        selectionJob?.cancel()
        if (film == null) {
            filmList.clear()
            resetColumnWidths()
            return
        }

        selectionJob = dialogScope.launch {
            val matchingFilms = withContext(Dispatchers.Default) {
                findDuplicateFilms(film)
            }
            filmList.clear()
            filmList.addAll(matchingFilms)
            calculateColumnWidths()
        }
    }

    private fun findDuplicateFilms(film: DatenFilm): List<DatenFilm> {
        val normalUrl = film.urlNormalQuality
        val highQualityUrl = film.highQualityUrl
        return Daten.getInstance().listeFilme.snapshot()
            .asSequence()
            .filter { item -> !item.isLivestream }
            .filter { item ->
                item.urlNormalQuality == normalUrl &&
                    item.highQualityUrl == highQualityUrl
            }
            .toList()
    }

    private fun calculateColumnWidths() {
        table.columnModel.columns.asIterator().forEach { column ->
            var preferredWidth = column.minWidth
            val columnIndex = column.modelIndex
            val headerRenderer = column.headerRenderer ?: table.tableHeader.defaultRenderer
            val header = headerRenderer.getTableCellRendererComponent(
                table,
                column.headerValue,
                false,
                false,
                0,
                columnIndex,
            )
            val maxWidth = header.preferredSize.width

            for (row in 0 until table.rowCount) {
                val cellRenderer = table.getCellRenderer(row, columnIndex)
                val component = table.prepareRenderer(cellRenderer, row, columnIndex)
                val width = component.preferredSize.width + table.intercellSpacing.width
                preferredWidth = maxOf(preferredWidth, width)

                if (preferredWidth <= maxWidth) {
                    preferredWidth = maxWidth
                    break
                }
            }
            column.preferredWidth = preferredWidth
        }

        table.doLayout()
    }

    private fun resetColumnWidths() {
        table.columnModel.columns.asIterator()
            .forEach { column -> column.preferredWidth = DEFAULT_COLUMN_WIDTH }
        table.doLayout()
    }

    private fun compareSenders(first: String, second: String): Int {
        val firstPenalty = senderPenalty(first)
        val secondPenalty = senderPenalty(second)
        if (firstPenalty != secondPenalty) {
            return firstPenalty.compareTo(secondPenalty)
        }
        return first.compareTo(second)
    }

    private fun senderPenalty(sender: String): Int =
        if (sender in PENALIZED_SENDERS) 1 else 0

    private companion object {
        private const val ROOT_NODE_LABEL = "Filmduplikate"
        private const val DEFAULT_COLUMN_WIDTH = 90
        private val PENALIZED_SENDERS = setOf("ARD", "ZDF")
    }
}
